File: 백준/Gold/16964. DFS 스페셜 저지/DFS 스페셜 저지.java
import java.util.*;
import java.lang.*;
import java.io.*;

class Main {
    static int n;
    static ArrayList<Integer>[] graph;
    static int[] priority, order;
    
	public static void main(String[] args) throws IOException {
	    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	    StringTokenizer st;
	    n = Integer.parseInt(br.readLine());
	    
	    graph = new ArrayList[n+1];
	    
	    for(int i = 1; i <= n; i++){
	        graph[i] = new ArrayList<>();
	    }
	    
	    for(int i = 0; i < n-1; i++){
	        st = new StringTokenizer(br.readLine());
	        int a = Integer.parseInt(st.nextToken());
	        int b = Integer.parseInt(st.nextToken());
	        
	        graph[a].add(b);
	        graph[b].add(a);
	    }
	    
	    priority = new int[n+1];
	    order = new int[n+1];
	    
	    st = new StringTokenizer(br.readLine());
	   
	    for(int i = 0; i < n; i++){
	        int a = Integer.parseInt(st.nextToken());
	        priority[a] = i;
	        order[i] = a;
	    }
	    
	    if(order[0] != 1){
	        System.out.println(0);
	    }
	    else{
	        for(int i = 1; i <= n; i++){
	            Collections.sort(graph[i], (a, b) -> Integer.compare(priority[a], priority[b]));
	        }
	        
	        if(dfs(1)){
	            System.out.println(1);
	        }
	        else{
	            System.out.println(0);
	        }
	    }
	}
	
	static boolean dfs(int cur){
	    Stack<Integer> st = new Stack<>();
	    boolean[] visited = new boolean[n+1];
	    
	    List<Integer> actual = new ArrayList<>();
	    st.add(1);
	    visited[1] = true;
	    
	    while(!st.isEmpty()){
	        int x = st.pop();
	        actual.add(x);
	        
	        for(int i = graph[x].size() - 1; i >= 0; i--){
	            if(!visited[graph[x].get(i)]){
	                st.add(graph[x].get(i));
	                visited[graph[x].get(i)] = true;
	            }
	        }
	    }
	    
	    for(int i = 0; i < n; i++){
	        if(actual.get(i) != order[i])
	            return false;
	    }
	    return true;
	}
}

File: 백준/Gold/16964. DFS 스페셜 저지/DFS 스페셜 저지.cc
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

vector<int> graph[100001], order, v;
int n, priority[100001], visited[100001];

bool cmp(const int &a, const int &b){
    return priority[a] < priority[b];
}

void dfs(int cur){
    if(visited[cur])
        return;
    
    visited[cur] = 1;
    v.push_back(cur);
    
    for(int i = 0; i < graph[cur].size(); i++){
        dfs(graph[cur][i]);   
    }
}

int main() {
    cin.tie(NULL);
    ios::sync_with_stdio(false);
    cin >> n;
    int a, b;
    for(int i = 0; i < n-1; i++){
        cin >> a >> b;
        graph[a].push_back(b);
        graph[b].push_back(a);
    }
    
    for(int i = 0 ; i < n; i++){
        cin >> a;
        order.push_back(a);
        priority[a] = i;
    }
    
    for(int i = 1; i <= n; i++){
        sort(graph[i].begin(), graph[i].end(), cmp);
    }
    
    dfs(1);
    
    int flag = 1;
    for(int i = 0; i < n; i++){
       if(v[i] != order[i]){
           flag = 0;
           break;
       }
    }
    
    cout << flag;
    return 0;
}